File: PERSONA/PERSONA.py
from abc import ABC, abstractmethod
#importo la libreria de metodo abstracto

class Persona(ABC):
    def __init__(self, nombre: str, apellidoPat: str, apellidoMat: str, cal: float, matri: str):
        self.nombre = nombre
        self.apellidoPat = apellidoPat
        self.apellidoMat = apellidoMat
        self.cal = cal
        self.matri = matri
#declaro la clase persona que va a ser mi arreglo de tipo abstracto con los elementos de nombres apellidos y de mas

    @abstractmethod
    def mostrar_info(self) -> str:
        pass

class Estudiante(Persona):
    def mostrar_info(self) -> str:
        return f"{self.nombre} {self.apellidoPat} {self.apellidoMat} {self.cal} {self.matri}"
#aqui asigno una clase hija que hereda los atributos de la clase persona y defino una funcion que sea mostrar la informacion que yo escriba para cada elemento

# Crear lista de 2 estudiantes
students = [None] * 2

#aqui le asigno los datos a cada estudiante usando la clase estudiante
students[0] = Estudiante("Jesus Alfonso", "Castro", "Sepulveda", 9.9, "24133450")
students[1] = Estudiante("Naomi", "Fong", "Burgueño", 9.9, "24125140")

#y aqui imprimo cada estudiante en un ciclo para no escribir linea por linea print
for i in range(len(students)):
    print(students[i].mostrar_info())

File: PERSONA/PERSONA.cpp
#include <iostream>
#include <string>
#include <vector>
using namespace std;

// Clase abstracta
class Persona {
protected:
    string nombre;
    string apellidoPat;
    string apellidoMat;
    float cal;
    string matri;

public:
    // Constructor
    Persona(string n, string ap, string am, float c, string m)
        : nombre(n), apellidoPat(ap), apellidoMat(am), cal(c), matri(m) {}

    // Método abstracto
    virtual void mostrar_info() const = 0;

    // Destructor virtual
    virtual ~Persona() {}
};

// Clase derivada
class Estudiante : public Persona {
public:
    // Constructor
    Estudiante(string n, string ap, string am, float c, string m)
        : Persona(n, ap, am, c, m) {}

    void mostrar_info() const override {
        cout << nombre << " " << apellidoPat << " " << apellidoMat
             << " " << cal << " " << matri << endl;
    }
};

int main() {
    vector<Persona*> students;

    students.push_back(new Estudiante("Jesus Alfonso", "Castro", "Sepulveda", 9.9, "24133450"));
    students.push_back(new Estudiante("Naomi", "Fong", "Burgueño", 9.9, "24125140"));

    for (const auto& student : students) {
        student->mostrar_info();
    }

    // Liberar memoria
    for (auto student : students) {
        delete student;
    }

    return 0;
}
